aict_tools/io.py

```diff
@@ -263,91 +387,159 @@ def has_holes(values):
+            if columns:
+                if "azimuth" in columns or "altitude" in columns:
+                    tel_key = tel.split('/')[-1]
+                    tel_triggers = pd.read_hdf(
+                        path,
+                        "/dl1/event/telescope/trigger"
+                    )
+                    tel_df = tel_df.merge(
+                        tel_triggers,
+                        how='inner',
+                        on=['obs_id', 'event_id', 'tel_id']
+                    )
+                    tel_pointings = pd.read_hdf(
+                        path,
+                        f"/dl1/monitoring/telescope/pointing/{tel_key}"
+                    )
+                    tel_df = tel_df.merge(
+                        tel_pointings,
+                        how='left',
+                        on='time'
+                    )
+                    # for chunked reading there might not be a matching trigger time
+                    if tel_df['azimuth'].isnull().iloc[0]:
+                        if aict_config.datamodel_version > '1.0.0': # is this the correct version?
+                            time_key = "time"
+                        else:
+                            time_key = "telescopetrigger_time"
+                        # find the closest earlier pointing
+                        earliest_chunktime = tel_df[time_key].min()
+                        time_diff = (
+                            tel_pointings['time']
+                            - earliest_chunktime
+                        )
+                        earlier = (time_diff <= 0)
+                        closest_pointing = tel_pointings.loc[time_diff[earlier].idxmax()]
+                        tel_df.at[0, 'azimuth'] = closest_pointing['azimuth']
+                        tel_df.at[0, 'altitude'] = closest_pointing['altitude']
+                    tel_df[['azimuth', 'altitude']] = tel_df[['azimuth', 'altitude']].fillna(
+                        method='ffill'
+                    )
+
+                if "equivalent_focal_length" in columns:
+                    tel_df = tel_df.merge(layout[['tel_id', 'equivalent_focal_length']], how='left', on='tel_id')
+
+            # combine the telescope dataframes
+            tel_dfs.append(tel_df)
+
+        # Monte carlo information is located in the simulation group
+        # and we are interested in the array wise true information only
+        df = pd.concat(tel_dfs)
+        if columns:
+            true_columns = [x for x in columns if x.startswith("true")]
+            if true_columns:
+                true_information = pd.read_hdf(
+                    path,
+                    "/simulation/event/subarray/shower")[
+                        true_columns+['obs_id', 'event_id']
+                    ]
+                df = df.merge(true_information, on=['obs_id', 'event_id'], how="left")
+
+
+        # this is super hacky and intransparent!
```

Comment:
Why? Dropping all non-needed columns is the only solution here. It is not hacky and certainly expected.

Probably add a `.copy` so the full df can be garbage collected.

Reply:
Also, dropping the columns before merging the tables would be better for memory consumption if we use less columns than are in the files.